### PR TITLE
Hyperparameter tuning

### LINKED ISSUE #21: Continue Fine-Tuning CodeT5 with R Bioinformatics Dataset and Evaluate Effectiveness
Continue with the fine-tuning process of the CodeT5 model using the expanded dataset of R bioinformatics code from sources like Bioconductor, tutorials, and DataScience Meta.

Document all the processes about the effectiveness of the process and what can be improved

Acceptance Criteria:
- Complete the next phase of fine-tuning using the updated dataset.
- Document observations, improvements, and areas needing more data or adjustment.

### PR DESCRIPTION
**Pull request by PranaviKolouju**
Fixes #21 

**What was changed?**

I created a jupyter notebook that contains a simple training script which trains on four different learning rates to find the most optimal one for our model finetuning (hyper parameter tuning the learning rate). Additionally, I added a markdown file that documents the exact learning rates tested as well as how each of them performed. 

**Why was it changed?**

In order to finetune our model, we need to train over the most optimal hyperparameters for our training objective. The script I wrote allows us to carefully adjust our learning rate parameters to one that will lead to model converge when later running it on our whole training dataset. Finding the optimal learning rate is particularly important because it ensures that we aren't erasing the generalizability of the model while also ensuring that the model is appropriately learning the nuances of our data. 

**How was it changed?**

I added the following files: babysitting-trials.ipynb and babysitting hyperparameter tuning.md. In the babysitting-trials notebook, I have implemented a simple training setup to tune the learning rate parameter. I created a dummy training set and masked the inputs similarly to how we masked our training data for the code completion/inpainting objective. I used the babysitting methodology, which says that an optimal learning rate will allow the model to converge within the first few epochs of training using just a handful of datapoints. After running this script, I found that 5e-4 was the most optimal learning rate for our purpose. I additionally added a markdown file to document what the babysitting methodology is and our finding from implementing this idea.

### CODE REVIEW

File: Fine Tuning/babysitting-trials.ipynb [hunk 1]
```diff
@@ -0,0 +1,877 @@
...
+    "        save_strategy=\"epoch\",\n",
+    "        load_best_model_at_end=True,\n",
+    "        logging_steps=1,\n",
+    "    )\n",
+    "\n",
+    "    class DummyDataset(torch.utils.data.Dataset):\n",
+    "        def __init__(self, encodings):\n",
+    "            self.encodings = encodings\n",
+    "\n",
+    "        def __getitem__(self, idx):\n",
```
Comment (oam67) [inline Fine Tuning/babysitting-trials.ipynb:827]: I noticed that you are cloning and detaching tensors inside __getitem__. What is the purpose of doing both? I wonder if the dataset isn't being modified after creation then would we need both? A direct slice might work as an alternative!

File: Fine Tuning/babysitting-trials.ipynb [hunk 2]
```diff
@@ -0,0 +1,877 @@
...
+     "output_type": "stream",
+     "text": [
+      "{'eval_loss': 23.869091033935547, 'eval_runtime': 0.5815, 'eval_samples_per_second': 6.879, 'eval_steps_per_second': 1.72, 'epoch': 5.0}\n"
+     ]
+    },
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "There were missing keys in the checkpoint model loaded: ['encoder.embed_tokens.weight', 'decoder.embed_tokens.weight', 'lm_head.weight'].\n",
```
Comment (oam67) [inline Fine Tuning/babysitting-trials.ipynb:238]: I noticed the repeated warning that mentioned 'encoder.embed_tokens.weight' and others missing during checkpoint loading. Is the model being reloaded properly each time? Could this be related to the masked token setup in some way?

File: Fine Tuning/babysitting-trials.ipynb [hunk 3]
```diff
@@ -0,0 +1,877 @@
...
+    "    )\n",
+    "\n",
+    "    trainer.train()\n",
+    "\n",
+    "    evaluation_results = trainer.evaluate(eval_dataset)\n",
+    "    print(f\"Evaluation results for LR {lr}: {evaluation_results}\")\n",
+    "\n",
+    "    print(f\"Final training loss for LR {lr}: {trainer.state.log_history[-1]}\")\n",
+    "\n",
+    "    model = AutoModelForSeq2SeqLM.from_pretrained(model_name)"
```
Comment (oam67) [inline Fine Tuning/babysitting-trials.ipynb:852]: It is good that you are reloading the model after each learning rate test. This is likely to help ensure results are not affected by previous training runs.

File: Fine Tuning/hyperparameter tuning.md [hunk 4]
```diff
@@ -0,0 +1,8 @@
+# Hyperparameter Tuning
+## Learning Rate
+
+I made a simple finetuning script to tune the learning rate parameter. According to the babysitting methodology, we can test whether a learning rate will lead the training to converge by simple training it over a handful of datapoints. An optimal learning rate should allow the model to converge over the first few epochs. 
+
+# Methods and Results
+
+I constructed a simple training framework that runs for 5 epochs, testing the following learning rates commonly used in finetuning transformer architectures: 1e-5, 5e-5, 1e-4, and 5e-4. It was observed that learning rates of 1e-5 and 5e-5 are too small and will likely prevent the model from converging optimally. On the other hand, a learning rate of 1e-4 proved to be quite unstable. 5e-4 was the most optimal learning rate of the batch and will likely be the learning rate that we can use to further  finetune our model with the final script.
```
Comment (oam67) [inline Fine Tuning/hyperparameter tuning.md:8]: It is really good that you have identified that 5e-4 will likely be the learning rate that we can use to further fine-tune our model. This will save us some time in doing hyperparameter tuning later on.

**Review by AjithAkuthota23 — APPROVED** [review]
Good work
